accepted/00000-use-dnf-for-rpm-repos.md

```diff
@@ -0,0 +1,177 @@
+- Provide contact & download parameters to the Downloader Thread for downloading the packages.
+- Resolve package dependencies. Since `spacewalk-repo-sync` allows filtering (include/exclude) packages when syncing, package dependencies need to be calculated when running on this mode.
+
+With this information collected by the plugin, the `DownloaderThread` from `reposync`
+is able to perform the actual download of the necessary package files when `spacewalk-repo-sync` is called.
+
+Currently, `spacewalk-repo-sync` comes with the following plugins:
+- `yum_src`: RPM based repositories
+- `deb_src`: DEB based repositories
+- `uln_src`: Oracle ULN repositories (a wrapper of `yum_src` with custom authentication)
+
+### The "ContentSource" plugin API
+
+Mandatory methods:
+- `get_md_checksum_type`: Return the checksum_type of primary.xml
+- `get_products`: Return products metadata if any
+- `get_susedata`: Return suse metadata if any
+- `get_updates`: Return "updateinfo" / "patches" info
+- `get_groups`: Return repo groups if any
+- `get_file`:
+- `get_modules`: Return module metadata if any
+- `raw_list_packages`: Return the raw list of packages available after filtering.
+- `list_packages`: Return a list of `ContentPackage` rhn class for the available packages after filtering.
+- `clear_cache`: Clear cache files
+- `set_download_parameters`: Prepare the downloader params dictionary
+
+Yum specific methods we also need to cover:
+- `repomd_up_to_date`: Check if repomd.xml checksum is up-to-date.
+- `get_metadata_paths`: Simply load primary and updateinfo path from repomd.
+
+#### The "ContentPackage" class
+A "ContentSource" plugin needs to provide a list of "ContentPackage" objects to the reposync. This RHN class represents the package like by:
+
+- NEVRA
+- unique ID
+- checksum
+- checksum type
+
+### Replacing the custom "Zypper" + "libsolv" plugin with a "DNF" plugin.
+
+The idea here would be the following:
+
+- The repository metadata downloading is delegated to DNF (which uses [librepo](http://rpm-software-management.github.io/librepo/) for the actual fetching)
+- The downloaded XML metadata can be manually read (packages, updates, products, etc).
+- Authentication with ULN repositories can be done by porting the YUM plugin to DNF
+- Legacy metadata not normally handled by DNF could be fetched by [librepo](http://rpm-software-management.github.io/librepo/) or urlgrabber through a plugin module like the ULN one
+
+#### How to calculate the dependencies?
+To access this information, we would use [the DNF API](https://dnf.readthedocs.io/en/latest/api.html) (OBS packages at [openSUSE:Factory](https://build.opensuse.org/package/show/openSUSE:Factory/dnf) - [openSUSE:Backports:SLE-15-SP1](https://build.opensuse.org/package/show/openSUSE:Backports:SLE-15-SP1/dnf)).
+
+#### Requirements for this approach:
+- We would need to use `python3-dnf` library ([OBS link](https://build.opensuse.org/package/show/openSUSE:Backports:SLE-15-SP1/dnf)) to do this.
```

Comment:
At least at this particular moment, we wouldn't like to add a new stack to maintain and support in our Enterprise products since we already have an alternative based on supported components.